src/cell_group.hpp

```diff
@@ -143,51 +143,61 @@ class cell_group {
     }
 
     void advance(time_type tfinal, time_type dt) {
-        while (lowered_.time()<tfinal) {
-            // take any pending samples
-            time_type cell_time = lowered_.time();
+        EXPECTS(lowered_.state_synchronized());
+
+        // Bin pending events and enqueue on lowered state.
+        time_type ev_min_time = lowered_.max_time(); // (but we're synchronized here)
+        while (auto ev = events_.pop_if_before(tfinal)) {
+            auto handle = get_target_handle(ev->target);
+            auto binned_ev_time = binner_.bin(ev->target.gid, ev->time, ev_min_time);
+            lowered_.add_event(binned_ev_time, handle, ev->weight);
+        }
+
+        lowered_.start_integration(tfinal, dt);
```

Comment:
The name is a bit misleading, maybe `prepare_integration` or `setup_integration` are more appropriate.

If I understand the comments and your plans correctly, the event delivery and sampler polling will both be moved to the lowered cells, and thus be able to perform time stepping to `t_final` with one call?

In that case, would we need to have a separate `start_integration` and `step_integration` methods?